src/compile/data/bin.ts

```diff
@@ -16,60 +16,59 @@ function numberFormatExpr(expr: string, format: string) {
   return `format(${expr}, '${format}')`;
 }
 
+function addRangeFormula(model: Model, transform: VgTransform[], fieldDef: FieldDef, channel: Channel) {
+    const discreteDomain = hasDiscreteDomain(model.scale(channel).type);
+    if  (discreteDomain) {
+      // read format from axis or legend, if there is no format then use config.numberFormat
+      const format = (model.axis(channel) || model.legend(channel) || {}).format ||
+        model.config.numberFormat;
+
+      const startField = field(fieldDef, {datum: true, binSuffix: 'start'});
+      const endField = field(fieldDef, {datum: true, binSuffix: 'end'});
+
+      transform.push({
+        type: 'formula',
+        as: field(fieldDef, {binSuffix: 'range'}),
+        expr: `${numberFormatExpr(startField, format)} + ' - ' + ${numberFormatExpr(endField, format)}`
+      });
+    }
+}
+
 function parse(model: Model): Dict<VgTransform[]> {
   return model.reduceFieldDef(function(binComponent: Dict<VgTransform[]>, fieldDef: FieldDef, channel: Channel) {
-    const bin = model.fieldDef(channel).bin;
-    if (bin) {
-
-       let binTrans: VgBinTransform = {
-        type: 'bin',
-        field: fieldDef.field,
-        as: [field(fieldDef, {binSuffix: 'start'}), field(fieldDef, {binSuffix: 'end'})],
-        signal: varName(model.getName(fieldDef.field + '_bins'))
-      };
-
-      if (!isBoolean(bin)) {
-         binTrans = extend(binTrans, bin);
+    const fieldDefBin = model.fieldDef(channel).bin;
+    if (fieldDefBin) {
+      const bin: Bin = isBoolean(fieldDefBin) ? {} : fieldDefBin;
+      if (!bin.maxbins && !bin.steps) {
```

Comment:
If we already initialize this in unit, why do we have to augment the value again here?